### ISSUE TITLE
Add STRICTENC flags to bitcoin::PublicKey::from_slice()?

### ISSUE THREAD

**Issue by panicfarm** [rust-bitcoin/rust-bitcoin#2333]
Some legacy uncompressed pubKeys did not have `0x04` prefix. For example, in the tx [`71d5c755efc891cd9273e4e889358d5f897f1e0e0e01dca7549c51d707f9fe49`](https://blockstream.info/tx/71d5c755efc891cd9273e4e889358d5f897f1e0e0e01dca7549c51d707f9fe49?expand) inp 0, the corresponding P2SH redeem script is 1 of 2 multisig, that has one compressed (length 33) and one uncompressed (length 65) pubkeys
```
1 0378d430274f8c5ec1321338151e9f27f4c676a008bdf8638d07c0b6be9ab35c71 0778d430274f8c5ec1321338151e9f27f4c676a008bdf8638d07c0b6be9ab35c71a1518063243acd4dfe96b66e3f2ec8013c8e072cd09b3834a19f81f659cc3455 2 OP_CHECKMULTISIG
```

It is the uncompressed 65 byte pubKey `0778d4..c3455` with `0x07` prefix that satisfies the input signature, but only if the flag [`SCRIPT_VERIFY_STRICTENC`]( https://github.com/bitcoin/bitcoin/blob/3ba8de1b704d590fa4e1975620bd21d830d11666/doc/release-notes/release-notes-0.10.0.md?plain=1#L482) is disabled. I verified this with `btcdeb`: it fails by default, but succeeds with 
```
btcdeb --modify-flags="-STRICTENC" --tx=010000000123d773a6dff771f9e32566b1d3fb08dcd9a0b2be78881461eef23e0861f2de3c01000000b2004730440220363939e550920b4d9947659a70f1c40603230c3e6967d20f22da130976ba01c1022056560d6f480b1bd150faf238c6ab36b42ad03d502970cce0a3092052ba5efe88014c6751210378d430274f8c5ec1321338151e9f27f4c676a008bdf8638d07c0b6be9ab35c71410778d430274f8c5ec1321338151e9f27f4c676a008bdf8638d07c0b6be9ab35c71a1518063243acd4dfe96b66e3f2ec8013c8e072cd09b3834a19f81f659cc345552aeffffffff01905f01000000000017a914367e3c2c31cb061606e5a812257fc153f2bef80e8700000000 --txin=0100000001990d1b71e3b572926982fa2419144aaccbc14ba2e918496052d26a4da3777132000000006a4730440220387b5e2d0005ff13f9a02df718fb85a67133484adc8e2571379aed1fccc0692b02204cc4555d33244e2abffcf362dcd4b71b5f373578c3a39da90bf120c2d3d80f9601210254df56fd5663610ce8370a8af0aa9e3e8bc1befd176554f35575769aec249d16ffffffff0280380100000000001976a914f679b8239ffdf39987d7c5ca8e5fd44475a795e188aca08601000000000017a9147e76c2468a8d453bfbc130e0ca86b6821b24c7b38700000000
```

I think we need a way to be able to add this flag to `bitcoin::PublicKey::from_slice()` call, because without it, it fails with `PubKey(InvalidKeyPrefix(7))` [here](https://github.com/rust-bitcoin/rust-bitcoin/blob/de9f20a62052224da824ddc101a838fd36020bd2/bitcoin/src/crypto/key.rs#L184).

**Comment by apoelstra** [comment]
cc #520 

I guess if some jerk actually used hybrid keys on-chain then we should support them somehow.

**Comment by Kixunil** [comment]
Yes. `PublicKey::from_slice_non_strict`? We then need to call it when parsing keys in `Script`. Though maybe we should start thinking about validation contexts...

**Comment by apoelstra** [comment]
@Kixunil we never parse keys in `Script` AFAIK. If we do that's very new and I can't imagine what it's used for.

We do parse keys in rust-miniscript, which already has facilities for enabling/disabling checks on parsing. Would be fine adding a mode to parse hybrid keys there.

**Comment by Kixunil** [comment]
[We do.](https://docs.rs/bitcoin/latest/bitcoin/blockdata/script/struct.Script.html#method.p2pk_public_key)

IIRC it has something to do with message signing.

**Comment by apoelstra** [comment]
Cute. We parse public keys to support *another* piece of deprecated cruft that has no address format and is therefore only possible to use by manually constructing transactions in a deliberately and pointlesssly inefficient way.

However, that specific method returns a `bitcoin::PublicKey` which does not have any way to indicate whether the key is hybrid (it only has a compressedness boolean). So we can't change it to support hybrid keys.

**Comment by Kixunil** [comment]
Oh, yeah, the address format is not supported for hybrid keys and thus the function is actually correct and since `PublicKey` doesn't mark whether it's hybrid we can't change it without adding the marker which I would love to avoid.

@panicfarm can you provide your code so that we can better understand your use case?

**Comment by Kixunil** [comment]
> _another_ piece of deprecated cruft that has no address format and is therefore only possible to use by manually constructing transactions in a deliberately and pointlesssly inefficient way.

I don't know what you're referring to.

**Comment by apoelstra** [comment]
@Kixunil p2pk outputs have no addresses and are not supported by basically any wallet software.

**Comment by apoelstra** [comment]
>The method should return tuple where the other element is bool indicating whether the key is hybrid.

No, we should absolutely not change the API so that every user needs to know what a "hybrid key" is.

**Comment by Kixunil** [comment]
That API is only used by people who already process P2PK so they should. Nobody else has to.

**Comment by apoelstra** [comment]
A very long time ago, Bitcoin Core supported p2pk outputs with compressed or uncompressed keys. (Actually I am uncertain about compressed keys, whether those predate p2pkh or not.) No wallet in the history of Bitcoin has ever supported hybrid keys.

Because actual wallets used p2pk outputs it is sometimes useful for software to be able to parse them, e.g. when trying to recover very old coins. I agree that the p2pk API is obscure and pretty-much useless but it's not "parsing an arbitrary crypto object that we found in the openSSL API and it appears somebody put on-chain somewhere" useless.

The only reason to support hybrid keys is if you are parsing obscure scripts and want to detect hybrid keys. But there is a practically infinite set of weird obscure objects you might want to detect in old scripts.

**Comment by apoelstra** [comment]
We could add another API function `p2pk_maybe_hybrid` or something which returns a secp pubkey and an enum covering all the possibilities. (BTW there are two kinds of hybrid keys, "hybrid even" and "hybrid odd", so a boolean wouldn't be sufficient anyway.)

**Comment by Kixunil** [comment]
> We could add another API function `p2pk_maybe_hybrid`

I had this idea too but I was worried that someone might introduce a subtle bug by missing hybrid keys. However as you note:

> but it's not "parsing an arbitrary crypto object that we found in the openSSL API and it appears somebody put on-chain somewhere" useless.

I think you have a point. Whoever is doing some sort of chain validation (not that I would recommend trying) will at least have a test go through all existing blocks and uncover the problem. Whoever wants to recover coins because they are on stupid UTXO will get punished for creating it in the first place which is only fair. Anyone else won't lose money because of such bug.

So I guess it's fine.

**Comment by panicfarm** [comment]
> @panicfarm can you provide your code so that we can better understand your use case?

@Kixunil The use-case is fairly trivial: the code re-verifies all legacy tx inputs by feeding them into [ `verify_with_flags` ](https://github.com/rust-bitcoin/rust-bitcoinconsensus/blob/b8eb0e8fd6462e8f945aaca4f5485de96586c2ef/src/lib.rs#L113 )and then prints and stores the verified pubKey hashes. We had to fork and modify the C++ core dependency in  [`bitcoinconsensus`](https://github.com/rust-bitcoin/rust-bitcoinconsensus) to be able to return the successfully verified keys into Rust.

```rust
let pk = bitcoin::PublicKey::from_slice(&pubkey)?;  
println!("inp verified pk {:x}", pk.inner);         
                                                    
let verified_pkhash = pk.pubkey_hash();             
```
